### ISSUE TITLE
"A method named '0' is already defined" when calling Meteor.methods with non-object argument

### ISSUE THREAD

**Issue by dandv** [meteor/meteor#4447]
Open the console, and run `Meteor.method('foo')` twice.

```
Meteor.methods('foo', 1);
undefined
Meteor.methods('foo', 1);
eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42 Uncaught Error: A method named '0' is already defined
    at https://d14jjfgstdxsoz.cloudfront.net/eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42:13985
    at Function.A.each.A.forEach (https://d14jjfgstdxsoz.cloudfront.net/eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:1:928)
    at a.extend.methods (https://d14jjfgstdxsoz.cloudfront.net/eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42:13934)
    at <anonymous>:2:8
    at Object.InjectedScript._evaluateOn (<anonymous>:895:140)
    at Object.InjectedScript._evaluateAndWrap (<anonymous>:828:34)
    at Object.InjectedScript.evaluate (<anonymous>:694:21)(anonymous function) @ eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42A.each.A.forEach @ eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:1a.extend.methods @ eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42(anonymous function) @ VM1578:2InjectedScript._evaluateOn @ VM1431:895InjectedScript._evaluateAndWrap @ VM1431:828InjectedScript.evaluate @ VM1431:694
Meteor.methods('bar', 1);
eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42 Uncaught Error: A method named '0' is already defined
    at https://d14jjfgstdxsoz.cloudfront.net/eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42:13985
    at Function.A.each.A.forEach (https://d14jjfgstdxsoz.cloudfront.net/eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:1:928)
    at a.extend.methods (https://d14jjfgstdxsoz.cloudfront.net/eda3c85f0c3aeafb6fe5eb9b05e0696b06d6762c.js:42:13934)
    at <anonymous>:2:8
    at Object.InjectedScript._evaluateOn (<anonymous>:895:140)
    at Object.InjectedScript._evaluateAndWrap (<anonymous>:828:34)
    at Object.InjectedScript.evaluate (<anonymous>:694:21)
```

Would help those new to Meteor if there was an explicit error that an object needs to be passed.

**Comment by dandv** [comment]
Updated issue.

**Comment by ekatek** [comment]
Thanks for catching this. You are right -- crashing with the 'a method named 0' error is kind of unhelpful. We might fix this at some point, but pull requests are encouraged.

**Comment by BraveKenny** [comment]
Easy fix could be to check if an object is passed:
```
methods: function (methods) {
    var self = this;
    if (typeof methods !== 'object')
      throw new Error("Dictionnary of methods must be an object");
    _.each(methods, function (func, name) {
      if (self.method_handlers[name])
        throw new Error("A method named '" + name + "' is already defined");
      self.method_handlers[name] = func;
    });
  }
```
But hey, people may want to send an array for some reason, and who are we to judge. Plus, you can still give incorrect data such as `{foo: 'bar'}`. So I'm thinking checking for the function type:
```
methods: function (methods) {
    var self = this;
    _.each(methods, function (func, name) {
      if (typeof fund !== 'function')
        throw new Error("Method '" + name + "' must be a function");
      if (self.method_handlers[name])
        throw new Error("A method named '" + name + "' is already defined");
      self.method_handlers[name] = func;
    });
  }
```
That way you get an error on the first call. It is still not that explicit in the case of `'foo'`:
```
Uncaught Error: Method '0' must be a function
```
But in my opinion, it gives a decent clue, with the added benefit of covering a broader scope of mistakes at the same number of lines added.

**Comment by BraveKenny** [comment]
Pull  #4473 was merged to fix this issue.

**Comment by pgAdmin** [comment]
@SylvainBigonneau I think this issue can be closed #4473 already merged

**Comment by abernix** [comment]
@pgAdmin Thanks for pointing that out.

Closing this as it should be resolved by #4473 (f6f5020128d6f6debcca1eeca5fca8dc1c671de0) and released in Meteor 1.2 and later.
